src/Compilers/Core/Portable/Declarations/INamespaceOrTypeDeclaration.cs

```diff
@@ -0,0 +1,36 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Represents either a namespace or a type declaration.
+    /// </summary>
+    /// <remarks>
+    /// This interface is reserved for implementation by its associated APIs. We reserve the right to
+    /// change it in the future.
+    /// </remarks>
+    public interface INamespaceOrTypeDeclaration
+    {
+        /// <summary>
+        /// Get the declaration name.
+        /// </summary>
+        string Name { get; }
+
+        /// <summary>
+        /// Get all the children namespace and type declarations of this declaration.
+        /// </summary>
+        ImmutableArray<INamespaceOrTypeDeclaration> Children { get; }
+
+        /// <summary>
+        /// Returns true if this declaration is a namespace. If it is not a namespace, it must be a type.
+        /// </summary>
+        bool IsNamespace { get; }
+
+        /// <summary>
+        /// Returns true if this declaration is a type. If it is not a type, it must be a namespace.
+        /// </summary>
+        bool IsType { get; }
+    }
```

Comment:
while sufficient for the needs of completion, i really think we should also expose `ImmutableArray<Location> IDeclaration.NameLocations` for IDeclaration, and probably `ImmutableArray<SyntaxReference> SyntaxReferences` as well.  People are going to enumerate these and then ask the obvious question of "where is this, and how do i get to it?".

Reply:
Yes, you mentioned this in the other PR, and I agree with you that this is useful. But that can come later, I'd prefer to wait until some feature actually needs it.

Reply:
I'll defer... but i personally think that's a mistake here.  As a public API, i'd prefer we not just make it for your single case :)  I'll let compiler decide what they think there.